### ISSUE TITLE
multi-stage: npm i --save-dev – RangeError: Maximum call stack size exceeded

### ISSUE THREAD

**Issue by stefanpenner** [npm/npm#8138]
given a large dep tree attempting to `--save-dev` a dependency blows the stack during `loadExtraneous`.

```
git clone https://github.com/ember-cli/ember-cli
cd ember-cli
% npm install --save-dev es6-promise
undefined:0¬
(null)¬
RangeError: Maximum call stack size exceeded¬
```

current npm commit: #b2dd728
```
± % npm version                                                                                                                 !10102
{ 'ember-cli': '0.2.3',
  npm: '3.0.0-pre.7',
  http_parser: '2.3',
  modules: '14',
  node: '0.12.2',
  openssl: '1.0.1m',
  uv: '1.4.2-node1',
  v8: '3.28.73',
  zlib: '1.2.8' }
```

**Comment by stefanpenner** [comment]
This actually can be replicated several ways, another way is to remove a large dependency from package.json and attempt an `npm i`. In both situations it appears some operation on a large dep-graph either is in error, should needs to be re-implemented without language level recursion.

I suppose it would be nice if we already had tail recursion.

**Comment by othiym23** [comment]
Hmmm...

```
bauchelain% git clone https://github.com/ember-cli/ember-cli
Cloning into 'ember-cli'...
remote: Counting objects: 23090, done.
remote: Compressing objects: 100% (118/118), done.
remote: Total 23090 (delta 52), reused 0 (delta 0), pack-reused 22972
Receiving objects: 100% (23090/23090), 6.43 MiB | 323.00 KiB/s, done.
Resolving deltas: 100% (13993/13993), done.
Checking connectivity... done.
bauchelain% cd ember-cli
bauchelain% npm install --save-dev es6-promise
ember-cli@0.2.3 /Users/ogd/Documents/projects/ember-cli
└── es6-promise@2.1.1

bauchelain% npm version
{ 'ember-cli': '0.2.3',
  npm: '3.0.0-pre.7',
  ares: '1.10.0-DEV',
  http_parser: '2.3.0',
  modules: '43',
  node: '1.8.1',
  openssl: '1.0.2a',
  uv: '1.4.2',
  v8: '4.1.0.27',
  zlib: '1.2.8' }
```

The release version of `npm@3` absolutely has to work with Node 0.12, obviously, but do you still see this failure if you switch to io.js 1.8.1? I'm going to switch to Node 0.12.2 and see if I see the same failure, but I've tried a couple different times with the same git version of npm as you without seeing this failure.

**Comment by othiym23** [comment]
I just tried with Node 0.12.2, similarly without a failure:

```
bauchelain% rm -rf ember-cli
bauchelain% git clone https://github.com/ember-cli/ember-cli
Cloning into 'ember-cli'...
remote: Counting objects: 23090, done.
remote: Compressing objects: 100% (118/118), done.
remote: Total 23090 (delta 52), reused 0 (delta 0), pack-reused 22972
Receiving objects: 100% (23090/23090), 6.43 MiB | 435.00 KiB/s, done.
Resolving deltas: 100% (13993/13993), done.
Checking connectivity... done.
bauchelain% cd ember-cli
bauchelain% nave use 0.12.2 node ../npm/npm/bin/npm-cli.js install --save-dev es6-promise
Already installed: 0.12.2
using 0.12.2
ember-cli@0.2.3 /Users/ogd/Documents/projects/ember-cli
└── es6-promise@2.1.1

bauchelain% nave use 0.12.2 node ../npm/npm/bin/npm-cli.js version
Already installed: 0.12.2
using 0.12.2
{ 'ember-cli': '0.2.3',
  npm: '3.0.0-pre.7',
  http_parser: '2.3',
  modules: '14',
  node: '0.12.2',
  openssl: '1.0.1m',
  uv: '1.4.2-node1',
  v8: '3.28.73',
  zlib: '1.2.8' }
```

**Comment by othiym23** [comment]
Can you gist me the output of running that same command with `-ddd` included?

Also (and I hope this isn't relevant), how much RAM is there in the system you're using?

**Comment by stefanpenner** [comment]
io.js appears fine, node 0.12.2 appears to choke, but not always... 

```

stefanpenner@trustdream-lm ~/tmp/broapp [master *]
± % node -v                                                                                                                                                                                                                              !10137
v0.12.2

stefanpenner@trustdream-lm ~/tmp/broapp [master *]
± % npm install --save rsvp                                                                                                                                                                                                              !10149
loadExtraneous            ▐ ╢███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░╟

undefined:0
(null)

RangeError: Maximum call stack size exceeded

stefanpenner@trustdream-lm ~/tmp/broapp [master *]
± % npm install --save rsvp                                                                                                                                                                                                              !10150
loadExtraneous            ▄ ╢███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░╟

undefined:0
(null)

RangeError: Maximum call stack size exceeded

stefanpenner@trustdream-lm ~/tmp/broapp [master *]
± % npm install --save-dev rsvp                                                                                                                                                                                                          !10150
loadExtraneous            ▌ ╢███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░╟

undefined:0
(null)

RangeError: Maximum call stack size exceeded

stefanpenner@trustdream-lm ~/tmp/broapp [master *]
± % nave use 1.7.1                                                                                                                                                                                                                       !10151
Already installed: 1.7.1
using 1.7.1
Victorious warriors win first and then go to war, while defeated warriors go to war first and then seek to win.
Warning! PATH is not properly set up, '/Users/stefanpenner/.rvm/gems/ruby-2.2.2/bin' is not at first place,
         usually this is caused by shell initialization files - check them for 'PATH=...' entries,
         it might also help to re-add RVM to your dotfiles: 'rvm get stable --auto-dotfiles',
         to fix temporarily in this shell session run: 'rvm use ruby-2.2.2'.
Victorious warriors win first and then go to war, while defeated warriors go to war first and then seek to win.

stefanpenner@trustdream-lm ~/tmp/broapp [master *]
± % npm install --save-dev rsvp                                                                                                                                                                                                          !10047
npm WARN package.json did_it_work@0.0.6 No repository field.
npm WARN package.json dom-serializer@0.1.0 No README data
npm WARN package.json es-simpler-traverser@0.0.1 No README data
npm WARN package.json object-component@0.0.3 No repository field.
npm WARN package.json parsejson@0.0.1 No repository field.
npm WARN package.json parsejson@0.0.1 No README data
npm WARN package.json parseqs@0.0.2 No repository field.
npm WARN package.json parseqs@0.0.2 No README data
npm WARN package.json parseuri@0.0.2 No repository field.
npm WARN package.json parseuri@0.0.2 No README data
npm WARN package.json styled_string@0.0.1 No repository field.
npm WARN package.json callsite@1.0.0 No repository field.
npm WARN package.json has-binary@0.1.5 No repository field.
npm WARN package.json has-binary-data@0.1.3 No repository field.
npm WARN package.json indexof@0.0.1 No repository field.
npm WARN package.json string_decoder@0.10.31 string_decoder is also the name of a node core module.
rsvp@3.0.18 node_modules/rsvp

stefanpenner@trustdream-lm ~/tmp/broapp [master *]
± % nave use 0.12.0                                                                                                                                                                                                                      !10048
Already installed: 0.12.0
using 0.12.0
Victorious warriors win first and then go to war, while defeated warriors go to war first and then seek to win.
rWarning! PATH is not properly set up, '/Users/stefanpenner/.rvm/gems/ruby-2.2.2/bin' is not at first place,
         usually this is caused by shell initialization files - check them for 'PATH=...' entries,
         it might also help to re-add RVM to your dotfiles: 'rvm get stable --auto-dotfiles',
         to fix temporarily in this shell session run: 'rvm use ruby-2.2.2'.
mVictorious warriors win first and then go to war, while defeated warriors go to war first and then seek to win.
 -%

stefanpenner@trustdream-lm ~/tmp/broapp [master *]
± % rm -rf node_modules/rsvp                                                                                                                                                                                                             !10049

stefanpenner@trustdream-lm ~/tmp/broapp [master *]
± % npm install --save-dev rsvp                                                                                                                                                                                                          !10050
npm WARN package.json did_it_work@0.0.6 No repository field.
npm WARN package.json dom-serializer@0.1.0 No README data
npm WARN package.json es-simpler-traverser@0.0.1 No README data
npm WARN package.json object-component@0.0.3 No repository field.
npm WARN package.json parsejson@0.0.1 No repository field.
npm WARN package.json parsejson@0.0.1 No README data
npm WARN package.json parseqs@0.0.2 No repository field.
npm WARN package.json parseqs@0.0.2 No README data
npm WARN package.json parseuri@0.0.2 No repository field.
npm WARN package.json parseuri@0.0.2 No README data
npm WARN package.json styled_string@0.0.1 No repository field.
npm WARN package.json callsite@1.0.0 No repository field.
npm WARN package.json has-binary@0.1.5 No repository field.
npm WARN package.json has-binary-data@0.1.3 No repository field.
npm WARN package.json indexof@0.0.1 No repository field.
npm WARN package.json string_decoder@0.10.31 string_decoder is also the name of a node core module.
rsvp@3.0.18 node_modules/rsvp

stefanpenner@trustdream-lm ~/tmp/broapp [master *]
± % node -v                                                                                                                                                                                                                              !10051
v0.12.0

stefanpenner@trustdream-lm ~/tmp/broapp [master *]
± %
```

**Comment by stefanpenner** [comment]
oh, wait im switching NPM versions when im switching node versions.

Give me a sec...

**Comment by stefanpenner** [comment]
failure with `-ddd` as requested:

stefanpenner@trustdream-lm ~/tmp/broapp [master *]
± % npm i -ddd --save rsvp                                                                                                                                                                                                               !10063
npm info it worked if it ends with ok
npm verb cli [ 'node',
npm verb cli   '/Users/stefanpenner/.nave/installed/0.12.0/bin/npm',
npm verb cli   'i',
npm verb cli   '-ddd',
npm verb cli   '--save',
npm verb cli   'rsvp' ]
npm info using npm@3.0.0-pre.7
npm info using node@v0.12.0
npm sill install loadCurrentTree
npm sill install readLocalPackageData
npm sill install normalizeTree
npm sill install loadIdealTree
npm sill install cloneCurrentTreeToIdealTree
npm sill install loadShrinkwrap
npm sill install loadAllDepsIntoIdealTree
npm sill fetchPackageMetaData rsvp@^3.0.18
npm sill fetchNamedPackageData rsvp
npm sill mapToRegistry name rsvp
npm sill mapToRegistry using default registry
npm sill mapToRegistry registry https://registry.npmjs.org/
npm sill mapToRegistry uri https://registry.npmjs.org/rsvp
npm verb request uri https://registry.npmjs.org/rsvp
npm verb request no auth needed
�npm info attempt registry request try #1 at 11:14:51 PM
npm verb request id 2912a62ba619a47e
npm verb etag "8RP80BWPRLEQ98TX8OMWXI6ZO"
npm http request GET https://registry.npmjs.org/rsvp
npm http 304 https://registry.npmjs.org/rsvp
npm sill get cb [ 304,
npm sill get   { date: 'Thu, 30 Apr 2015 06:14:50 GMT',
npm sill get     server: 'Apache',
npm sill get     via: '1.1 varnish',
npm sill get     'cache-control': 'max-age=60',
npm sill get     etag: '"8RP80BWPRLEQ98TX8OMWXI6ZO"',
npm sill get     age: '0',
npm sill get     'x-served-by': 'cache-sjc3121-SJC',
npm sill get     'x-cache': 'HIT',
npm sill get     'x-cache-hits': '1',
npm sill get     'x-timer': 'S1430374490.642800,VS0,VE86',
npm sill get     vary: 'Accept',
npm sill get     'content-length': '0',
npm sill get     'keep-alive': 'timeout=10, max=50',
npm sill get     connection: 'Keep-Alive' } ]
�npm verb etag https://registry.npmjs.org/rsvp from cache
�npm verb get saving rsvp to /Users/stefanpenner/.npm/registry.npmjs.org/rsvp/.cache.json
npm sill cache add args [ 'rsvp@^3.0.18', null ]
npm verb cache add spec rsvp@^3.0.18
�npm sill cache add parsed spec { raw: 'rsvp@^3.0.18',
npm sill cache add   scope: null,
npm sill cache add   name: 'rsvp',
npm sill cache add   rawSpec: '^3.0.18',
npm sill cache add   spec: '>=3.0.18 <4.0.0',
npm sill cache add   type: 'range' }
npm sill addNamed rsvp@>=3.0.18 <4.0.0
npm verb addNamed ">=3.0.18 <4.0.0" is a valid semver range for rsvp
npm sill addNameRange { name: 'rsvp', range: '>=3.0.18 <4.0.0', hasData: false }
npm sill mapToRegistry name rsvp
npm sill mapToRegistry using default registry
npm sill mapToRegistry registry https://registry.npmjs.org/
npm sill mapToRegistry uri https://registry.npmjs.org/rsvp
npm verb addNameRange registry:https://registry.npmjs.org/rsvp not in flight; fetching
npm verb get https://registry.npmjs.org/rsvp not expired, no request
npm sill addNameRange number 2 { name: 'rsvp', range: '>=3.0.18 <4.0.0', hasData: true }
npm sill addNameRange versions [ 'rsvp',
npm sill addNameRange   [ '1.0.0',
npm sill addNameRange     '1.1.0',
npm sill addNameRange     '1.1.1',
npm sill addNameRange     '1.2.0',
npm sill addNameRange     '2.0.0',
npm sill addNameRange     '2.0.1',
npm sill addNameRange     '2.0.2',
npm sill addNameRange     '2.0.3',
npm sill addNameRange     '2.0.4',
npm sill addNameRange     '3.0.0',
npm sill addNameRange     '3.0.1',
npm sill addNameRange     '3.0.2',
npm sill addNameRange     '3.0.3',
npm sill addNameRange     '3.0.4',
npm sill addNameRange     '3.0.6',
npm sill addNameRange     '3.0.7',
npm sill addNameRange     '3.0.8',
npm sill addNameRange     '3.0.9',
npm sill addNameRange     '3.0.12',
npm sill addNameRange     '3.0.14',
npm sill addNameRange     '3.0.13',
npm sill addNameRange     '3.0.15',
npm sill addNameRange     '3.0.16',
npm sill addNameRange     '3.0.17',
npm sill addNameRange     '3.0.18' ] ]
npm sill addNamed rsvp@3.0.18
npm verb addNamed "3.0.18" is a plain semver version for rsvp
npm sill cache afterAdd rsvp@3.0.18
npm verb afterAdd /Users/stefanpenner/.npm/rsvp/3.0.18/package/package.json not in flight; writing
npm verb afterAdd /Users/stefanpenner/.npm/rsvp/3.0.18/package/package.json written
npm sill addShrinkwrap Adding shrinkwrap to rsvp
loadExtraneous            ▀ ╢███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░╟

undefined:0
(null)

RangeError: Maximum call stack size exceeded

**Comment by stefanpenner** [comment]
Ah, I am a bit tired. My appologies, I just realized, I am not in `ember-cli` but in an app generated by ember-cli.

```
npm install -g ember-cli
ember new my-cool-app
cd my-cool-app
npm install --save-dev es6-promise
```

failures on both node 0.12.0 and iojs.

**Comment by othiym23** [comment]
I can reproduce, and will dig further.

**Comment by stefanpenner** [comment]
> I can reproduce, and will dig further.

:+1: thanks – flushing my own queue i'll circle back.

**Comment by iarna** [comment]
Fixed by #8138
